Feature/multi build

dsilhavy opened the pull request:
This PR addresses #4622.

## Description
As discussed in #4617 and #4619 developers have implemented different ways to include dash.js in their application. This PR applies changes to the Webpack configuration and the build process of dash.js to create two different bundles: 

A `legacy` bundle setting `ie: '11'` as the target for Babel and a `modern` bundle that uses a `.browserslistrc` file with target set to `defaults`. `defaults` is a shortcut for `is a shortcut for > 0.5%, last 2 versions, Firefox ESR, not dead.`. For both builds, we create a `UMD` and a `ESM` version. 

This PR contains multiple commits but the relevant one is this one: https://github.com/Dash-Industry-Forum/dash.js/commit/15a2bf1a61abf8f0f3e00a0ad5b2eb1ffd20c3f7. It includes:

* Adds a `.browserlistrc` file to define the target browsers for the build
* Moves all Webpack related configuration files to the `build/webpack` folder including subfolders for common configuration of the `legacy` and `modern` builds.
* Most of the code that is shared between the builds can be found in the `common` folder to avoid code duplication
* The final output files are now placed in subfolders within the `dist` folder. It contains a `legacy` folder and a `modern` folder with `esm` and `umd` subfolders for both. 
* In `webpack.legacy.base.cjs` the `'@babel/preset-env` target is set explicitly to `ie:11`. For unknown reasons, [ignoreBrowserslistConfig](https://babeljs.io/docs/babel-preset-env#ignorebrowserslistconfig) was not working for me.
* The `package.json` was adjusted to use the right files in `exports`. Moreover, `npm run build` now creates both the `legacy` and the `modern` build files. There are separate npm scripts to build only the `legacy` or the `modern` files.

The other commits in this PR adjust the path to the `dist` files accordingly and remove outdated samples.

## Open Questions
I am looking for any feedback people have on these changes. For instance:

* Do people agree on the general idea/approach to have two separate builds?
* Should we add all build files to the npm package? Or do we limit it to the `modern` build?
* Should we change the target setting for the `legacy`build. See also #4619 and #4617.
* Should we change the `.browserslistrc` config?
* Is the folder structure okay? Do we expect import problems when using subfolders inside `dist`?

## Open Todos
* Adjust the Github action (CI/CD) to run the right build command

littlespex commented on build/webpack/legacy/webpack.legacy.base.cjs:
```diff
@@ -0,0 +1,56 @@
+    },
+});
+
+const esmConfig = merge(legacyConfig, {
+    experiments: {
+        outputModule: true
+    },
+    output: {
+        path: path.resolve(__dirname, '../../../dist/legacy/esm'),
+        publicPath: '/dist/legacy/esm/',
+        library: {
+            type: 'module',
+        },
+        libraryExport: 'default',
+    },
+});
```
If `legacy` is essentially an ES5 build, how can there be an ESM version?

littlespex commented on build/webpack/modern/webpack.modern.base.cjs:
```diff
@@ -0,0 +1,53 @@
+    },
+},)
+
+const umdConfig = merge(modernConfig, {
+    output: {
+        path: path.resolve(__dirname, '../../../dist/modern/umd'),
+        publicPath: '/dist/modern/umd/',
+        library: 'dashjs',
+        libraryTarget: 'umd',
+        libraryExport: 'default'
+    },
+});
```
The `legacy` version is already a UMD bundle. What is the advantage of having a modern non-ESM variant? I would think that anyone needing to include the library via script/src would use legacy+umd and anyone in a modern browser, or using a build system, would use modern+esm.

littlespex reviewed (commented):
I would love to hear other opinions, but I think this could just be two bundles/folders:
- `dist/umd`: The ES5/UMD bundle that can be loaded via `<script src="" />` on older platforms (TVs mostly).
- `dist/esm`: The ESM bundler for modern browsers using `<script type="module" />` and app that use build tools to compile their JS.

bbert commented:
For information, the sample application `network-interceptor` provides an example of typescript application that can be used to validate importing of types and the modern+esm dash.js bundle.

dsilhavy commented:
Hi all,
thank you for the feedback and the contributions. Based on your comments, I did the following changes:

* Removed the `ESM` bundle from the `legacy` build 
* Removed the `core.js` polyfills from the `modern` builds
* Fixed the `npm run build` script to not execute `prebuild` twice
* Fixed the Github actions

That leaves us with the following builds:

* `UMD legacy`: A `UMD` build targeting legacy platforms by specifying the babel target  `ie: '11'`. In addition, `core.js` polyfills are enabled. 
* `ESM modern`: An `ESM` build using `.browserslistrc` as target, with target set to `defaults`. **No** `core.js` polyfills are enabled.
* `UMD modern`:  A `UMD` build targeting modern platforms using `.browserslistrc` as target, with target set to `defaults`. **No** `core.js` polyfills are enabled.

The  `UMD modern` build is up for discussion. From @littlespex comment above, I conclude that this is not really needed. However, you also mentioned that in many cases you add the polyfills yourself. Does it make sense to provide this additional UMD build without any polyfills?

Please also let me know if you see any other issues with these changes.

stschr commented:
No input from our side; all sounds reasonably here.
Well done!

eirikbjornr commented:
We'll likely continue pulling Dash.js' source files directly and building/bundling them ourself – just to remove a step in the chain of code transformations up to the build of our TV apps platform. We take a very similar approach to @story75 and share the same concerns as I describe in our previous discussion: https://github.com/Dash-Industry-Forum/dash.js/discussions/4617#discussioncomment-11508876

Our main concern is the polyfills needed to run Dash.js don't change drastically. Our tools will notify us if an unsupported polyfill is introduced so we're not flying blind here.

dsilhavy commented:
Thank you all for your input, I am merging this now. We can still apply changes based on what was done in this PR. I also adjusted the documentation to account for the new build formats: https://dashif.org/dash.js/pages/quickstart/installation.html
